### PR TITLE
feat(http_listener_v2): allows multiple paths and add path_tag

### PR DESCRIPTION
**Pull request by sumo-drosiek**
### Required for all PRs:

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.

resolves #

Convert `path` configuration to `paths` to allow getting metrics on multiple endpoints. In addition adding `path_tag`
 configuration option, so path can be included in defined tag name

### CODE REVIEW

File: plugins/inputs/http_listener_v2/README.md [hunk 1]
```diff
@@ -18,8 +18,12 @@ This is a sample configuration for the plugin.
   ## Address and port to host HTTP listener on
   service_address = ":8080"
 
-  ## Path to listen to.
-  # path = "/telegraf"
+  ## Paths to listen to.
+  # paths = ["/telegraf"]
```
Comment (srebhan) [inline plugins/inputs/http_listener_v2/README.md:22]: This is a breaking change, but we need to preserve backward compatibility for existing configurations. Thus, you need to keep the `path` option (maybe mark deprecated).

File: plugins/inputs/http_listener_v2/README.md [hunk 2]
```diff
@@ -18,8 +18,12 @@ This is a sample configuration for the plugin.
...
+
+  ## Save path in path_tag
+  ## Do not include path in tag if path_tag is an empty string
+  # path_tag = ""
```
Comment (srebhan) [inline plugins/inputs/http_listener_v2/README.md:26]: This should be a boolean.

File: plugins/inputs/http_listener_v2/http_listener_v2.go [hunk 3]
```diff
@@ -36,7 +36,8 @@ type TimeFunc func() time.Time
 // HTTPListenerV2 is an input plugin that collects external metrics sent via HTTP
 type HTTPListenerV2 struct {
 	ServiceAddress string            `toml:"service_address"`
-	Path           string            `toml:"path"`
```
Comment (srebhan) [inline plugins/inputs/http_listener_v2/http_listener_v2.go:39]: Again, you need to keep this and add it to your `paths` list to preserve backward compatibility.

File: plugins/inputs/http_listener_v2/http_listener_v2.go [hunk 4]
```diff
@@ -36,7 +36,8 @@ type TimeFunc func() time.Time
...
+	Paths          []string          `toml:"paths"`
+	PathTag        string            `toml:"path_tag"`
```
Comment (srebhan) [inline plugins/inputs/http_listener_v2/http_listener_v2.go:40]: Please change this to
```suggestion
	PathTag        bool            `toml:"path_tag"`
```

File: plugins/inputs/http_listener_v2/http_listener_v2.go [hunk 5]
```diff
@@ -186,10 +193,20 @@ func (h *HTTPListenerV2) Stop() {
...
+func (h *HTTPListenerV2) containsPath(needle string) bool {
+	for _, v := range h.Paths {
+		if v == needle {
+			return true
+		}
+	}
+
+	return false
+}
+
```
Comment (srebhan) [inline plugins/inputs/http_listener_v2/http_listener_v2.go:205]: Please directly use the functionality from `"github.com/influxdata/telegraf/internal/choice"` as
```golang
choice.Contains(needle, h.Paths)
```

File: plugins/inputs/http_listener_v2/http_listener_v2.go [hunk 6]
```diff
@@ -186,10 +193,20 @@ func (h *HTTPListenerV2) Stop() {
...
 func (h *HTTPListenerV2) ServeHTTP(res http.ResponseWriter, req *http.Request) {
 	handler := h.serveWrite
 
-	if req.URL.Path != h.Path {
+	if !h.containsPath(req.URL.Path) {
```
Comment (srebhan) [inline plugins/inputs/http_listener_v2/http_listener_v2.go:209]:
```suggestion
	if !choice.Contains(req.URL.Path, h.Paths) {
```

File: plugins/inputs/http_listener_v2/http_listener_v2.go [hunk 7]
```diff
@@ -251,6 +268,10 @@ func (h *HTTPListenerV2) serveWrite(res http.ResponseWriter, req *http.Request)
 			}
 		}
 
+		if h.PathTag != "" {
+			m.AddTag(h.PathTag, req.URL.Path)
+		}
```
Comment (srebhan) [inline plugins/inputs/http_listener_v2/http_listener_v2.go:273]: With `PathTag` being a `bool`:
```suggestion
		if h.PathTag {
			m.AddTag("path", req.URL.Path)
		}
```

File: plugins/inputs/http_listener_v2/http_listener_v2_test.go [hunk 8]
```diff
@@ -49,7 +49,7 @@ func newTestHTTPListenerV2() *HTTPListenerV2 {
 	listener := &HTTPListenerV2{
 		Log:            testutil.Logger{},
 		ServiceAddress: "localhost:0",
-		Path:           "/write",
+		Paths:          []string{"/write"},
```
Comment (srebhan) [inline plugins/inputs/http_listener_v2/http_listener_v2_test.go:52]: Please keep those tests unmodified to check backward compatibility.

**Review by srebhan — COMMENTED** [review]
Hey @sumo-drosiek thanks for this nice contribution! I added some minor comments in the code. Please keep the change backward compatible as users rely on the fact that a new telegraf still works with old config files.

File: plugins/inputs/http_listener_v2/README.md (hunk 2 above)
Comment (sumo-drosiek) [reply plugins/inputs/http_listener_v2/README.md:26]: My intention is to set `tag` key here, so no hardcoded tag name is being used

File: plugins/inputs/http_listener_v2/http_listener_v2_test.go (hunk 8 above)
Comment (sumo-drosiek) [reply plugins/inputs/http_listener_v2/http_listener_v2_test.go]: Tests restored to original state

File: plugins/inputs/http_listener_v2/README.md (hunk 1 above)
Comment (sumo-drosiek) [reply plugins/inputs/http_listener_v2/README.md]: I simply appends `path` to `paths` to keep backward compability

File: plugins/inputs/http_listener_v2/README.md (hunk 2 above)
Comment (srebhan) [reply plugins/inputs/http_listener_v2/README.md]: To be honest, I think this is overengineered. You can still change the tag name with processors later and it makes the option harder to understand for the user. So my preference is to have a hard-coded tag-name.

File: plugins/inputs/http_listener_v2/README.md [hunk 9]
```diff
@@ -19,8 +19,12 @@ This is a sample configuration for the plugin.
   service_address = ":8080"
 
   ## Path to listen to.
+  ## This is depracated and will be appended to paths
```
Comment (srebhan) [inline plugins/inputs/http_listener_v2/README.md:22]: Maybe
```suggestion
  ## This option is deprecated and only available for backward-compatibility. Please use `paths` instead.
```
is more clear.

File: plugins/inputs/http_listener_v2/http_listener_v2.go [hunk 10]
```diff
@@ -64,8 +67,16 @@ const sampleConfig = `
...
   ## Path to listen to.
+  ## This is depracated and will be appended to paths
   # path = "/telegraf"
 
+  ## Paths to listen to.
+  # paths = ["/telegraf"]
+
+  ## Save path in path_tag
+  ## Do not include path in tag if path_tag is an empty string
+  # path_tag = ""
```
Comment (srebhan) [inline plugins/inputs/http_listener_v2/http_listener_v2.go:78]: If you change this, please remember to also update the REAMDE.

File: plugins/inputs/http_listener_v2/README.md (hunk 2 above)
Comment (sumo-drosiek) [reply plugins/inputs/http_listener_v2/README.md]: ok, but what if `path` is already included into metric which comes to the input?

**Review by srebhan — COMMENTED** [review]
Nice, leaves the path-tag issue to resolve. Thanks for working on this!

File: plugins/inputs/http_listener_v2/README.md (hunk 2 above)
Comment (srebhan) [reply plugins/inputs/http_listener_v2/README.md]: This can only happen if the user configured it in `http_header_tags` I think. If you really worry about it, add a check if it was set there.

Comment (sumo-drosiek) [reply plugins/inputs/http_listener_v2/README.md]: I'm concern about parsers. For example as I understand, prometheusremotewrite treats all labels as tags: https://github.com/influxdata/telegraf/blob/32d4234ae4edb39a9ae86915d457974d167a0d0c/plugins/parsers/prometheusremotewrite/parser.go#L37-L39

Comment (srebhan) [reply plugins/inputs/http_listener_v2/README.md]: Let me rephrase: Which value would you use for `path_tag` in your setup? Why isn't this a nice name for the tag then? :-)

There is always the risk of a collision between that and the "path-tag". I'm not bound to the name `path`, so we could use something more specific (`listener_path`?). But imaging you offer that option and set `path_tag="some_value_path"`. How do you guarantee that the data you receive does not contain that tag?

So my idea is to _detect_ if the name is already used in the parsed metric and then use e.g. `http_listener_v2_path` as tag-name instead or issue a warning that the name is already taken. This will cover the whole collision problem and you will need it even if you allow setting a custom name.

Comment (sumo-drosiek) [reply plugins/inputs/http_listener_v2/README.md]: Ok, I get it now :)
`path` is pretty popular name and there is high chance to be used already by some of the metrics. Could we just use `http_listener_v2_path`? It can be pretty complicated to detect if `path` is from original source or from `http_listener_v2`.

Due to above I would like to avoid `path` in my setup

File: config.yaml [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+[[outputs.file]]
+  ## Files to write to, "stdout" is a specially handled file.
+  files = ["stdout"]
+
+  ## Data format to output.
+  ## Each data format has its own unique set of configuration options, read
+  ## more about them here:
+  ## https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_OUTPUT.md
+  data_format = "prometheus"
```
Comment (srebhan) [inline config.yaml:20]: Please remove this file.

File: plugins/inputs/http_listener_v2/http_listener_v2.go [hunk 12]
```diff
@@ -371,6 +389,7 @@ func init() {
 			ServiceAddress: ":8080",
 			TimeFunc:       time.Now,
 			Path:           "/telegraf",
+			Paths:          []string{"/telegraf"},
```
Comment (srebhan) [inline plugins/inputs/http_listener_v2/http_listener_v2.go:392]: You need to remove `Path` here as otherwise `Paths` has a duplicate entry.

File: plugins/inputs/http_listener_v2/http_listener_v2.go [hunk 13]
```diff
@@ -136,6 +147,9 @@ func (h *HTTPListenerV2) Start(acc telegraf.Accumulator) error {
 		h.WriteTimeout = config.Duration(time.Second * 10)
 	}
 
+	// Append h.Path to h.Paths
+	h.Paths = append(h.Paths, h.Path)
+
```
Comment (srebhan) [inline plugins/inputs/http_listener_v2/http_listener_v2.go:152]: Please only add `Path` if it is not empty and not yet contained in `h.Paths` as otherwise you might introduce duplicate entries.

**Review by srebhan — COMMENTED** [review]
Thanks for the quick cycles. The only thing left is to only add the `Path` if it is not empty and not yet in `Paths` (you create a duplication by default in your `init()`). Furthermore, there is a spurious `config.yaml` file that should not go into the repo I thinks.

File: plugins/inputs/http_listener_v2/http_listener_v2.go (hunk 12 above)
Comment (sumo-drosiek) [reply plugins/inputs/http_listener_v2/http_listener_v2.go:392]: You mean entire line (it requires to change tests) or just to set it to empty string?

Comment (sumo-drosiek) [reply plugins/inputs/http_listener_v2/http_listener_v2.go]: nevermind, fixed :)

**Comment by sumo-drosiek** [comment]
@srebhan I believe all issues has been resolved

**Review by srebhan — APPROVED** [review]
Looks good to me. Thanks for this nice contribution @sumo-drosiek!

**Review by sspaink — APPROVED** [review]
lgtm, thanks for the contribution!
